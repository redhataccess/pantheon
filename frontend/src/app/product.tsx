import React, { Component } from 'react';
import { Bullseye, Button, Alert, AlertActionCloseButton, Form, FormGroup, TextInput, ActionGroup } from '@patternfly/react-core';
import '@app/app.css';
import { Redirect } from 'react-router-dom'

class Product extends Component {
  public state = {
    failedPost: false,
    isDup: false,
    isMissingFields: false,
    login: false,
    productDescription: '',
    productName: '',
    redirect: false,
    results: [],
  };

  // render method transforms the react components into DOM nodes for the browser.
  public render() {
    const { productName, productDescription, isMissingFields, isDup } = this.state;
    return (
      <React.Fragment>
        {/* Bullseye makes sure everyhting is in the middle */}
        <Bullseye>
          <Form>
            <div className="app-container">
              <div>
                {isMissingFields && (
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="A Product name is required."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                )}
                {isDup && (
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="Duplicated Product name."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                )}
                <FormGroup
                  label="Product Name"
                  isRequired={true}
                  fieldId="product-name"
                >
                  <TextInput isRequired={true} id="product-name" type="text" placeholder="Product Name" value={productName} onChange={this.handleNameInput} />
                </FormGroup>
                <br />
                <FormGroup
                  label="Product Description"
                  fieldId="product-description"
                >
                  <TextInput id="product-description" type="text" placeholder="Product Description" value={productDescription} onChange={this.handleProductInput} />
                </FormGroup>
                <br />
                <ActionGroup>
                  <Button aria-label="Creates a new Product Name with Description specified." onClick={this.saveProduct}>Save</Button>
                  <div>
                    {this.checkAuth()}
                    {this.loginRedirect()}
                    {this.renderRedirect()}
                  </div>
                </ActionGroup>
              </div>
            </div>
          </Form>
        </Bullseye>
      </React.Fragment>
    );
  }
  // methods that handle the state changes.
  private handleNameInput = productName => {
    this.setState({ productName });
  };

  private handleProductInput = productDescription => {
    this.setState({ productDescription });
    // console.log("Desc " + productDescription)
  };

  private saveProduct = () => {
    if (this.state.productName === '') {
      this.setState({ isMissingFields: true })
    } else {
      this.productExist().then(exist => {
        if (!exist) {
          const hdrs = {
            'Accept': 'application/json',
            'cache-control': 'no-cache'
          }
          // setup url fragment
          const urlFragment = this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_");
          const formData = new FormData();
          formData.append("name", this.state.productName)
          formData.append("description", this.state.productDescription)
          formData.append("sling:resourceType", "pantheon/product")
          formData.append("jcr:primaryType", 'pant:product')
          // currently we don't translate products in Customer Portal.
          formData.append("locale", "en-US")
          formData.append("url", urlFragment)
          // fetch makes the request to create a new product.
          // transfor productName to lower case and replace special chars with _.
          fetch(encodeURI('/content/products/' + urlFragment), {
            body: formData,
            headers: hdrs,
            method: 'post'
          }).then(response => {
            if (response.status === 201 || response.status === 200) {
              this.setState({ redirect: true })
            } else if (response.status === 500) {
              // console.log(" Needs login " + response.status)
              this.setState({ login: true })
            } else {
              console.log(" Failed " + response.status)
              this.setState({ failedPost: true })
            }
          });
        }
      }
      )
    }
  }

  private renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/products' />
    } else {
      return ""
    }
  }

  private loginRedirect = () => {
    if (this.state.login) {
      return <Redirect to='/login' />
    } else {
      return ""
    }
  }

  private checkAuth = () => {
    fetch("/system/sling/info.sessionInfo.json")
      .then(response => response.json())
      .then(responseJSON => {
        const key = "userID"
        if (responseJSON[key] === 'anonymous') {
          this.setState({ login: true })
        }
      })
  }

  private dismissNotification = () => {
    this.setState({ isMissingFields: false, isDup: false });
  }

  private productExist = () => {
    let exists = false
    const backend = '/content/products/' + this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_") + '.json'
    return fetch(backend)
      .then(response => {
        if (response.status === 200) {
          exists = true
          this.setState({ isDup: true })
        }
      }).then(() => exists)
  }
}

export { Product }