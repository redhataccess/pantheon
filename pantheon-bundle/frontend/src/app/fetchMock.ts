if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}