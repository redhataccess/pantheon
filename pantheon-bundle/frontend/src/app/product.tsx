import React, { Component } from 'react'
import { Bullseye, Button, Alert, AlertActionCloseButton, Form, FormGroup, TextInput, ActionGroup } from '@patternfly/react-core'
import '@app/app.css'
import { Redirect } from 'react-router-dom'

interface IState {
  failedPost: boolean
  isDup: boolean
  isMissingFields: boolean
  productDescription: string
  productName: string
  redirect: boolean
}

class Product extends Component<any, IState> {
  constructor(props) {
    super(props)
    this.state = {
      failedPost: false,
      isDup: false,
      isMissingFields: false,
      productDescription: '',
      productName: '',
      redirect: false
    }

  }

  // render method transforms the react components into DOM nodes for the browser.
  public render() {
    return (
      <React.Fragment>
        {/* Bullseye makes sure everyhting is in the middle */}
        <Bullseye>
          <Form>
            <div className="app-container">
              <div>
                {this.state.isMissingFields && 
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="A Product name is required."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                }
                {this.state.isDup && 
                  <div className="notification-container">
                    <Alert
                      variant="warning"
                      title="Duplicated Product name."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    />
                  </div>
                }
                {this.state.failedPost &&
                  <div className="notification-container">
                    <Alert
                      variant="danger"
                      title="Failed to create product."
                      action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                    >
                    Please check if you are logged in as a publisher.
                    </Alert>
                  </div>
                }
                <FormGroup
                  label="Product Name"
                  isRequired={true}
                  fieldId="product-name"
                >
                  <TextInput isRequired={true} id="product-name" type="text" placeholder="Product Name" value={this.state.productName} onChange={this.handleNameInput} />
                </FormGroup>
                <br />
                <FormGroup
                  label="Product Description"
                  fieldId="product-description"
                >
                  <TextInput id="product-description" type="text" placeholder="Product Description" value={this.state.productDescription} onChange={this.handleProductInput} />
                </FormGroup>
                <br />
                <ActionGroup>
                  <Button aria-label="Creates a new Product Name with Description specified." onClick={this.saveProduct}>Save</Button>
                  <div>
                    {this.renderRedirect()}
                  </div>
                </ActionGroup>
              </div>
            </div>
          </Form>
        </Bullseye>
      </React.Fragment>
    )
  }
  // methods that handle the state changes.
  private handleNameInput = productName => {
    this.setState({ productName })
  }

  private handleProductInput = productDescription => {
    this.setState({ productDescription })
    // console.log("Desc " + productDescription)
  }

  private saveProduct = () => {
    if (this.state.productName === '') {
      this.setState({ isMissingFields: true })
    } else {
      this.productExist().then(exist => {
        if (!exist) {
          const hdrs = {
            'Accept': 'application/json',
            'cache-control': 'no-cache'
          }
          // setup url fragment
          const urlFragment = this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_")
          const formData = new FormData()
          formData.append("name", this.state.productName)
          formData.append("description", this.state.productDescription)
          formData.append("sling:resourceType", "pantheon/product")
          formData.append("jcr:primaryType", 'pant:product')
          // currently we don't translate products in Customer Portal.
          formData.append("locale", "en-US")
          formData.append("url", urlFragment)
          // fetch makes the request to create a new product.
          // transfor productName to lower case and replace special chars with _.
          fetch(encodeURI('/content/products/' + urlFragment), {
            body: formData,
            headers: hdrs,
            method: 'post'
          }).then(response => {
            if (response.status === 201 || response.status === 200) {
              this.setState({ redirect: true })
            } else {
              this.setState({ failedPost: true })
            }
          })
        }
      }
      )
    }
  }

  private renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/products' />
    } else {
      return ""
    }
  }

  private dismissNotification = () => {
    this.setState({ isMissingFields: false, isDup: false })
  }

  private productExist = () => {
    let exists = false
    const backend = '/content/products/' + this.state.productName.toString().toLowerCase().replace(/[^A-Z0-9]+/ig, "_") + '.json'
    return fetch(backend)
      .then(response => {
        if (response.status === 200) {
          exists = true
          this.setState({ isDup: true })
        }
      }).then(() => exists)
  }
}

export { Product }