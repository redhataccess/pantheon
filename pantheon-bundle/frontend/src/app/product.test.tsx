import React from "react"
import { Product } from "@app/product"
import "@app/fetchMock"

import { mount, shallow } from "enzyme"
import { TextInput, FormGroup, Button } from "@patternfly/react-core"
import { render, fireEvent } from '@testing-library/react'

describe("Product tests", () => {
  test("should render Product component", () => {
    const view = shallow(<Product />)
    expect(view).toMatchSnapshot()
  })

  it("should render a form group", () => {
    const wrapper = mount(<Product />)
    const formGroup = wrapper.find(FormGroup)
    expect(formGroup.exists()).toBe(true)
  })

  it("should render a text input", () => {
    const wrapper = mount(<Product />)
    const textInput = wrapper.find(TextInput)
    expect(textInput.exists()).toBe(true)
  })

  it("should render a Button", () => {
    const wrapper = mount(<Product />)
    const button = wrapper.find(Button)
    expect(button.exists()).toBe(true)
  })
 
  it("test handleNameInput function", () => {
    const container = shallow(<Product />)
    const productInput = container.find('#product_name_text')
    productInput.simulate('change', 'test_name');
    container.update()
    expect(container.find('#product_name_text').prop('value')).toEqual(
      'test_name',
    );
  })
  it("test handleProductInput function", () => {
    const container = shallow(<Product />)
    const input = container.find('#product_description_text')
    input.simulate('change', 'test_prod_descrip');
    container.update()
    expect(container.find('#product_description_text').prop('value')).toEqual(
      'test_prod_descrip',
    );
  })

  it("test handleUrlInput function - valid url provided", () => {
    const container = shallow(<Product />)
    const input = container.find('#product_url_fragment_text')
    input.simulate('change', 'test_prod_url');
    container.update()
    expect(container.find('#product_url_fragment_text').prop('value')).toEqual(
      'test_prod_url',
    );
  })
  it("test handleUrlInput function - invalid url provided", () => {
    const container = shallow(<Product />)
    const input = container.find('#product_url_fragment_text')
    input.simulate('change', '');
    container.update()
    expect(container.find('#product_url_fragment_text').prop('value')).toEqual(
      '',
    );
  })
  it("test handleTextInputChange function", () => {
    const container = shallow(<Product />)
    const input = container.find('#new_version_name_text')
    input.simulate('change', 'test_version_name');
    container.update()
    expect(container.find('#new_version_name_text').prop('value')).toEqual(
      'test_version_name',
    );
  })
  it("test handleUrlInputChange function - valid URL", () => {
    const container = shallow(<Product />)
    const input = container.find('#new_version_url_fragment')
    input.simulate('change', 'test_version_url');
    container.update()
    expect(container.find('#new_version_url_fragment').prop('value')).toEqual(
      'test_version_url',
    );
  })
  it("test handleUrlInputChange function - invalid URL", () => {
    const container = shallow(<Product />)
    const input = container.find('#new_version_url_fragment')
    input.simulate('change', '');
    container.update()
    expect(container.find('#new_version_url_fragment').prop('value')).toEqual(
      '',
    );
  })
  test('test alert message that appears when form is missing fields', () => {
    const handleClose = jest.fn()
    const { queryByText, getByText, getByLabelText } = render(<Product onClose={handleClose} />)
    fireEvent.click(getByText('Save'))
    expect(queryByText("Fields indicated by * are mandatory")).toBeTruthy()
    const closeButton = getByLabelText('Close Warning alert: alert: Fields indicated by * are mandatory')
    fireEvent.click(closeButton)

  })
  test('test alert message that appears when invalid input is entered for URL Fragment', () => {
    const { queryByText, getByPlaceholderText } = render(<Product />)
    let input = getByPlaceholderText('URL Fragment')
    fireEvent.change(input, { target: { value: '*' } })
    expect(queryByText("Allowed input for Product ulrFragment: alphanumeric, hyphen, period and underscore")).toBeTruthy()
  })
  test('test filling out product form and clicking save product button', async () => {
    const { getByText, getByPlaceholderText } = render(<Product exist={false} />)
    let productInput = getByPlaceholderText('Product Name')
    fireEvent.change(productInput, { target: { value: 'test_name' } })
    let urlInput = getByPlaceholderText('URL Fragment')
    fireEvent.change(urlInput, { target: { value: 'url' } })
    let prodVersionInput = getByPlaceholderText('Product Version')
    fireEvent.change(prodVersionInput, { target: { value: 'test_version' } })
    let versionUrlInput = getByPlaceholderText('Version URL Fragment')
    fireEvent.change(versionUrlInput, { target: { value: 'test_url' } })
    fireEvent.click(getByText('Save'))
  })
})
