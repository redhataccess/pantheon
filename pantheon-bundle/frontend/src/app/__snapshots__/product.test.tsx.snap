// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product tests should render Product component 1`] = `
<Fragment>
  <Bullseye>
    <Form>
      <div
        className="app-container"
      >
        <div>
          <FormGroup
            fieldId="product-name"
            isRequired={true}
            label="Product Name"
          >
            <ForwardRef
              id="product-name"
              isRequired={true}
              onChange={[Function]}
              placeholder="Product Name"
              type="text"
              value=""
            />
          </FormGroup>
          <br />
          <FormGroup
            fieldId="product-description"
            label="Product Description"
          >
            <ForwardRef
              id="product-description"
              onChange={[Function]}
              placeholder="Product Description"
              type="text"
              value=""
            />
          </FormGroup>
          <br />
          <ActionGroup>
            <Component
              aria-label="Creates a new Product Name with Description specified."
              onClick={[Function]}
            >
              Save
            </Component>
            <div />
          </ActionGroup>
        </div>
      </div>
    </Form>
  </Bullseye>
</Fragment>
`;

exports[`Product tests test checkAuth function 1`] = `undefined`;

exports[`Product tests test dismissNotification function 1`] = `[Function]`;

exports[`Product tests test handleNameInput function 1`] = `undefined`;

exports[`Product tests test handleProductInput function 1`] = `undefined`;

exports[`Product tests test loginRedirect function 1`] = `undefined`;

exports[`Product tests test productExist function 1`] = `Promise {}`;

exports[`Product tests test renderRedirect function 1`] = `[Function]`;
